### PR TITLE
add RoFormer Model

### PR DESCRIPTION
**Pull request by JunnYu**
### Description

添加RoFormer模型

### CODE REVIEW

**Review by ZeyuChen — COMMENTED** [review]
Please also add an example of RoFormer to paddlenlp's examples. eg.

https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/language_model/roformer

And also please provide more pretrained weight sent to @yingyibiao, we will help you upload the pretrained weight of RoFromer.

**Comment by JunnYu** [comment]
@yingyibiao
麻烦添加一下有关的roformer权重，谢谢
https://drive.google.com/file/d/1-j_gCJ3oYjiGMG6ISfnrhQIzoe3h0bwH/view

**Comment by yingyibiao** [comment]
> @yingyibiao
> 麻烦添加一下有关的roformer权重，谢谢
> https://drive.google.com/file/d/1-j_gCJ3oYjiGMG6ISfnrhQIzoe3h0bwH/view

done

File: paddlenlp/transformers/roformer/modeling.py [hunk 1]
```diff
@@ -0,0 +1,779 @@
...
+            "pad_token_id": 0,
+            "rotary_value": True,
+        },
+    }
+
+    resource_files_names = {"model_state": "model_state.pdparams"}
+    pretrained_resource_files_map = {
+        "model_state": {
+            "roformer-chinese-small":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-small/model_state.pdparams",
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/modeling.py]: modify model_state url:
"https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-small/model_state.pdparams"
->
"https://paddlenlp.bj.bcebos.com/models/transformers/roformer/roformer-chinese-small/model_state.pdparams"

File: paddlenlp/transformers/roformer/__init__.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+from .modeling import *
+from .tokenizer import *
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/__init__.py]: add a new line at the end of file

File: paddlenlp/transformers/roformer/modeling.py [hunk 3]
```diff
@@ -0,0 +1,779 @@
...
+            "roformer-chinese-sim-char-ft-base":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-sim-char-ft-base/model_state.pdparams",
+            "roformer-chinese-sim-char-small":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-sim-char-small/model_state.pdparams",
+            "roformer-chinese-sim-char-base":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-sim-char-base/model_state.pdparams",
+            "roformer-english-small-discriminator":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-english-small-discriminator/model_state.pdparams",
+            "roformer-english-small-generator":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-english-small-generator/model_state.pdparams",
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/modeling.py]: fix the rest urls

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+            # the following line get: '今天 的 天气 非常 好 ！'
+            tokenizer.convert_tokens_to_string(tokens)
+    """
+
+    resource_files_names = {"vocab_file": "vocab.txt"}  # for save_pretrained
+    pretrained_resource_files_map = {
+        "vocab_file": {
+            # chinese word level model
+            "roformer-chinese-small":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-small/vocab.txt",
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: modify vocab url:
"https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-small/vocab.txt"
->
"https://paddlenlp.bj.bcebos.com/models/transformers/roformer/roformer-chinese-small/vocab.txt"

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-sim-char-ft-base/vocab.txt",
+            "roformer-chinese-sim-char-small":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-sim-char-small/vocab.txt",
+            "roformer-chinese-sim-char-base":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-chinese-sim-char-base/vocab.txt",
+            # english
+            "roformer-english-small-discriminator":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-english-small-discriminator/vocab.txt",
+            "roformer-english-small-generator":
+            "https://huggingface.co/junnyu/roformer_paddle/resolve/main/roformer-english-small-generator/vocab.txt",
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: fix the rest urls

Comment (JunnYu) [reply paddlenlp/transformers/roformer/tokenizer.py]: done all

**Comment by yingyibiao** [comment]
please install pre-commit to format code style e.g.: pip install pre-commit & pre-commit install

**Comment by JunnYu** [comment]
> please install pre-commit to format code style e.g.: pip install pre-commit & pre-commit install

@yingyibiao 
你好，我pip install pre-commit & pre-commit install ，然后提交的时候发现没反应
image

**Comment by yingyibiao** [comment]
> > please install pre-commit to format code style e.g.: pip install pre-commit & pre-commit install
> 
> @yingyibiao
> 你好，我pip install pre-commit & pre-commit install ，然后提交的时候发现没反应
> image

可以对每个文件做一些格式上的小修改，然后重新提交～

**Review by yingyibiao — COMMENTED** [review]
converter这个目录可以删除，这个是遗留的老代码

File: paddlenlp/transformers/roformer/modeling.py [hunk 6]
```diff
@@ -0,0 +1,777 @@
...
+    "RoFormerPretrainingHeads",
+    "RoFormerForSequenceClassification",
+    "RoFormerForTokenClassification",
+    "RoFormerForQuestionAnswering",
+]
+
+dtype_float = paddle.get_default_dtype()
+
+
+class MultiHeadAttentionWithRotary(MultiHeadAttention):
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/modeling.py]: MultiHeadAttentionWithRotary这个类建议直接继承nn.Layer进行实现

File: paddlenlp/transformers/roformer/modeling.py [hunk 7]
```diff
@@ -0,0 +1,777 @@
...
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+from paddle import tensor
+from paddle.fluid import layers
+from paddle.nn import Layer
+from paddle.nn.layer.transformer import (
+    MultiHeadAttention,
+    _convert_attention_mask,
+    _convert_param_attr_to_list, )
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/modeling.py]: 这里不建议import内部方法

File: paddlenlp/transformers/roformer/modeling.py [hunk 8]
```diff
@@ -0,0 +1,777 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+from paddle import tensor
+from paddle.fluid import layers
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/modeling.py]: paddle.fluid 将在未来废弃，建议使用其他可替代的API。

**Comment by JunnYu** [comment]
@yingyibiao 
done
image

File: paddlenlp/transformers/roformer/modeling.py [hunk 9]
```diff
@@ -0,0 +1,777 @@
...
+    def __init__(self,
+                 embed_dim,
+                 num_heads,
+                 dropout=0.,
+                 kdim=None,
+                 vdim=None,
+                 need_weights=False,
+                 weight_attr=None,
+                 bias_attr=None,
+                 rotary_value=False):
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/modeling.py:59]: 建议将rotary_value加入config

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+
+__all__ = ["RoFormerTokenizer"]
+
+
+class JiebaBasicTokenizer(object):
+    """
+    Runs basic tokenization (punctuation splitting, lower casing, etc.).
+    Args:
+        do_lower_case (bool): Whether the text strips accents and convert to
+            lower case. If you use the BERT Pretrained model, lower is set to
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: BERT -> Roformer

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+            sep_token="[SEP]",
+            pad_token="[PAD]",
+            cls_token="[CLS]",
+            mask_token="[MASK]", ):
+
+        if not os.path.isfile(vocab_file):
+            raise ValueError(
+                "Can't find a vocabulary file at path '{}'. To load the "
+                "vocabulary from a pretrained model please use "
+                "`tokenizer = BertTokenizer.from_pretrained(PRETRAINED_MODEL_NAME)`"
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: BERT -> Roformer

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+        """
+        return the size of vocabulary.
+        Returns:
+            int: the size of vocabulary.
+        """
+        return len(self.vocab)
+
+    def _tokenize(self, text, use_jieba=True):
+        """
+        End-to-end tokenization for BERT models.
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: BERT -> Roformer

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,524 @@
...
+        split_tokens = []
+        for token in self.basic_tokenizer.tokenize(text):
+            for sub_token in self.wordpiece_tokenizer.tokenize(token):
+                split_tokens.append(sub_token)
+
+        return split_tokens
+
+    def tokenize(self, text):
+        """
+        End-to-end tokenization for BERT models.
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: BERT -> Roformer

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+        return len(
+            self.build_inputs_with_special_tokens(token_ids_0, token_ids_1
+                                                  if pair else None))
+
+    def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None):
+        """
+        Build model inputs from a sequence or a pair of sequence for sequence classification tasks by concatenating and
+        adding special tokens.
+
+        A BERT sequence has the following format:
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: BERT -> Roformer

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 15]
```diff
@@ -0,0 +1,524 @@
...
+        return [(0, 0)] + offset_mapping_0 + [(0, 0)
+                                              ] + offset_mapping_1 + [(0, 0)]
+
+    def create_token_type_ids_from_sequences(self,
+                                             token_ids_0,
+                                             token_ids_1=None):
+        """
+        Create a mask from the two sequences passed to be used in a sequence-pair classification task.
+
+        A BERT sequence pair mask has the following format:
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: BERT -> Roformer

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 16]
```diff
@@ -0,0 +1,524 @@
...
+        _sep = [self.sep_token_id]
+        return _cls + token_ids_0 + _sep + token_ids_1 + _sep
+
+    def build_offset_mapping_with_special_tokens(self,
+                                                 offset_mapping_0,
+                                                 offset_mapping_1=None):
+        """
+        Build offset map from a pair of offset map by concatenating and adding offsets of special tokens.
+
+        A BERT offset_mapping has the following format:
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: BERT -> Roformer

File: paddlenlp/transformers/roformer/modeling.py (hunk 6 above)
Comment (JunnYu) [reply paddlenlp/transformers/roformer/modeling.py]: TransformerEncoderLayerWithRotary 这个也要重新定义嘛？

File: paddlenlp/transformers/roformer/tokenizer.py [hunk 17]
```diff
@@ -0,0 +1,524 @@
...
+            "do_lower_case": True
+        },
+        "roformer-chinese-sim-char-base": {
+            "do_lower_case": True
+        },
+        "roformer-english-small-discriminator": {
+            "do_lower_case": True
+        },
+        "roformer-english-small-generator": {
+            "do_lower_case": True
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformer/tokenizer.py]: 建议参考Albert将中文模型和英文模型的tokenizer统一

File: paddlenlp/transformers/roformer/modeling.py (hunk 6 above)
Comment (yingyibiao) [reply paddlenlp/transformers/roformer/modeling.py]: > TransformerEncoderLayerWithRotary 这个也要重新定义嘛？

这个可以不用

**Review by yingyibiao — COMMENTED** [review]
建议增加一个论文上的CAIL2019-SCM任务的example

File: paddlenlp/transformers/roformer/modeling.py (hunk 9 above)
Comment (JunnYu) [reply paddlenlp/transformers/roformer/modeling.py:59]: 这rotary_value在config中存在

File: paddlenlp/transformers/roformer/tokenizer.py (hunk 17 above)
Comment (JunnYu) [reply paddlenlp/transformers/roformer/tokenizer.py]: 我这里准备额外添加一个参数`use_jieba`来区分是否使用jieba。
`use_jieba =True`：word level的（中文）模型需要使用jieba，
`use_jieba =False`：char level的（中/英文）模型不需要使用jieba，此时roformer tokenizer等价于bert tokenizer。

File: paddlenlp/transformers/roformer/modeling.py (hunk 9 above)
Comment (yingyibiao) [reply paddlenlp/transformers/roformer/modeling.py:59]: ok

File: examples/language_model/roformer/README.md [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+3. 例子：
+   假设我想转换https://huggingface.co/junnyu/roformer_chinese_base 权重
+   - (1)首先下载 https://huggingface.co/junnyu/roformer_chinese_base/tree/main 中的pytorch_model.bin文件,假设我们存入了`./roformer_chinese_base/pytorch_model.bin`
+   - (2)运行convert.py
+        ```bash
+        python convert.py \
+            --pytorch_checkpoint_path ./roformer_chinese_base/pytorch_model.bin \
+            --paddle_dump_path ./roformer_chinese_base/model_state.pdparams
+        ```
+   - (3)最终我们得到了转化好的权重`./roformer_chinese_base/model_state.pdparams`
```
Comment (yingyibiao) [inline examples/language_model/roformer/README.md]: 这里可以略去，convert.py代码保留即可

File: examples/language_model/roformer/README.md [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+    --num_train_epochs 3 \
+    --logging_steps 1 \
+    --save_steps 500 \
+    --output_dir ./thucnews/ \
+    --device gpu \
+    --use_amp False
+```
+其中参数释义如下：
+- `model_type` 指示了模型类型，可以选择roformer。
+- `model_name_or_path` 指示了某种特定配置的模型，对应有其预训练模型和预训练时使用的 tokenizer。若模型相关内容保存在本地，这里也可以提供相应目录地址。注：`roformer-chinese-base`等对应使用的预训练模型转自[huggingface/transformers](https://github.com/huggingface/transformers)，具体可参考当前目录下`convert.py`中的内容。
```
Comment (yingyibiao) [inline examples/language_model/roformer/README.md]: 我们的权重已经上传了，所以这里不需要介绍如何转换了

File: examples/language_model/roformer/README.md [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+    --logging_steps 60 \
+    --save_steps 600 \
+    --output_dir ./cail2019_scm/ \
+    --device gpu \
+    --use_amp False
+```
+
+其中参数释义如下：
+- `model_type` 指示了模型类型，可以选择roformer_cls_pooling和roformer_mean_pooling两种类型。
+- `model_name_or_path` 指示了某种特定配置的模型，对应有其预训练模型和预训练时使用的 tokenizer。若模型相关内容保存在本地，这里也可以提供相应目录地址。注：`roformer-chinese-base`等对应使用的预训练模型转自[huggingface/transformers](https://github.com/huggingface/transformers)，具体可参考当前目录下`convert.py`中的内容。
```
Comment (yingyibiao) [inline examples/language_model/roformer/README.md]: 同上

File: examples/language_model/roformer/run_thucnews.py [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+    print("eval loss: %f, acc: %s, " % (loss.numpy(), res), end='')
+    model.train()
+
+
+def convert_example(example,
+                    tokenizer,
+                    label_list,
+                    max_seq_length=512,
+                    is_test=False):
+    """convert a glue example into necessary features"""
```
Comment (yingyibiao) [inline examples/language_model/roformer/run_thucnews.py]: glue -> thucnews

File: examples/language_model/roformer/README.md [hunk 22]
```diff
@@ -0,0 +1,128 @@
+# RoFormer
+
+## 模型简介
+
+[RoFormer](https://arxiv.org/pdf/2104.09864.pdf) (RoFormer: Enhanced Transformer with Rotary Position Embedding)是一个带有旋转位置嵌入(RoPE)的MLM预训练语言模型。 RoPE是一种相对位置编码方法，具有良好的理论特性。其主要思想是根据绝对位置将上下文嵌入（transformer中的 q，k）乘以旋转矩阵。可以证明上下文嵌入的内积将仅取决于相对位置。
+RoPE 是唯一可用于线性注意力的相对位置嵌入。更多详情请参考[论文](https://arxiv.org/pdf/2104.09864.pdf)或[原博客](https://kexue.fm/archives/8265)。EleutherAI还发布了一篇[博客](https://blog.eleuther.ai/rotary-embeddings/)，其中包含有关 RoPE 的直观解释和实验。
+
+本项目是RoFormer在 Paddle 2.0上的开源实现，包含了`THUCNews分类任务`和`Cail2019 Scm任务`的微调代码。
```
Comment (yingyibiao) [inline examples/language_model/roformer/README.md]: Paddle 2.0 -> Paddle 2.x

File: paddlenlp/transformers/roformer/tokenizer.py (hunk 17 above)
Comment (yingyibiao) [reply paddlenlp/transformers/roformer/tokenizer.py]: great！

**Review by yingyibiao — COMMENTED** [review]
Great job!

**Review by yingyibiao — APPROVED** [review]
LGTM

**Pull request merged**